freqtrade/freqai/data_drawer.py

```diff
@@ -263,23 +263,55 @@ def set_pair_dict_info(self, metadata: dict) -> None:
             self.pair_dict[metadata["pair"]] = self.empty_pair_dict.copy()
             return
 
-    def set_initial_return_values(self, pair: str, pred_df: DataFrame) -> None:
+    def set_initial_return_values(self, pair: str,
+                                  pred_df: DataFrame,
+                                  dataframe: DataFrame
+                                  ) -> None:
         """
         Set the initial return values to the historical predictions dataframe. This avoids needing
         to repredict on historical candles, and also stores historical predictions despite
         retrainings (so stored predictions are true predictions, not just inferencing on trained
-        data)
+        data).
+
+        We also aim to keep the date from historical predictions so that the FreqUI displays
+        zeros during any downtime (between FreqAI reloads).
         """
 
-        hist_df = self.historic_predictions
-        len_diff = len(hist_df[pair].index) - len(pred_df.index)
-        if len_diff < 0:
-            df_concat = pd.concat([pred_df.iloc[:abs(len_diff)], hist_df[pair]],
-                                  ignore_index=True, keys=hist_df[pair].keys())
+        new_pred = pred_df.copy()
+        # set new_pred values to nans (we want to signal to user that there was nothing
+        # historically made during downtime. The newest pred will get appeneded later in
+        # append_model_predictions)
+        new_pred.iloc[:, :] = np.nan
+        new_pred["date"] = dataframe["date"]
+
+        hist_preds = self.historic_predictions[pair].copy()
```

Comment:
In/after this line - what's the difference between `hist_preds` and `pred_df`? 

In my current understanding understanding, `set_initial_historic_predictions()` is assigning that same `pred_df` dataframe to exactly this dict, which effectively means at this point we have 3 copies dataframes of length X,
one empty, and 2 with identical (but copied) contents.

---

(leaving the above aside for now) - is it necessary to do a `.copy()` here?
`self.historic_predictions[pair]` is assigned below  - so i assume the prior dataframe is outdated, so mutating that would not matter, anyway (were not in "multi-threading" land here yet as far as i can tell?)

Reply:
All of FreqAI is multithreaded. We don't want simultaneous access to this while it has a changed column name.

Reply:
`pred_df` is a single vector of predictions coming out of the model.

`historic_predictions` contains everything that is appended and sent back to strat. See `append_model_predictions()` function just below.

Reply:
> All of FreqAI is multithreaded. We don't want simultaneous access to this while it has a changed column name.

This part / function is running in the main thread though (called via `start()` -> `start_live()` - which starts the thread, but then continues to this function) - which is why i'm asking.

I think i got the confusion about `pred_df` now - it's the odd code segment in `set_initial_historic_predictions()` (see link below) - which messes with references in an odd way - and should probably be cleaned up to be clearer.

https://github.com/freqtrade/freqtrade/blob/628963c20704ca7d47f4d2ec79cfaaf048fc8c82/freqtrade/freqai/freqai_interface.py#L655-L656

Reply:
I can see where the confusion would originate.

So we have the following events in FreqAI which are designed to accommodate FreqUI viewing.

First time a model is created and inferenced, we inference on the full set of exchange candles. We save those as "predictions" (but all except the last one are technically fake predictions). We set those as the first `historic_predictions` dataframe for that pair. We never do this fake prediction thing again after the first call. The goal is simply to fill out FreqUI as a starting point for users.

From then on, we only append to historic_predictions and then return the `tail` of that which is equivalent in length to the strategy dataframe (the exchange candles). 

The current PR is about reloading FreqAI, we need to load in those historic predictions, but we need to identify how many candles we were "offline for" so we can show in FreqUI that there were no predictions duringthose offline candles, and so that future candles are placed properly in FreqUI.

Reply:
well i'm not confused, i'm just pointing out that this is running in the main thread, which contradicts your above statement of "all of freqAI is multithreaded".


The code i highlighed above is clearly not clean code i'd like you to fix - but i understand you'd like to look at this at a later point / different PR :+1:

Reply:
We can chang it now, now problem. But the reason for the alias was actually to reduce line lengths and make the function easier to  read, all while manipulating the original df inside `self.historic_predictions[pair]`. I don’t believe we can assign pred_df to self.historic and then continue using pred_df below with the expectation that it’s mutating self.historic. 

Looking forward to your comment.

Reply:
```
 self.dd.historic_predictions[pair] = pred_df 
 hist_preds_df = self.dd.historic_predictions[pair] 
```

what the above does is assign `pred_df` to `self.dd.historic_predictions[pair]` - and then `self.dd.historic_predictions[pair]` to `hist_preds_df`, which creates an alias.
It's however not creating a copy or new reference (at no point in these 2 lines) - it's just storing a reference to pred_df in a dictionary (eventually adding a new key to that dictionary) - and then creating a new alias called `hist_preds_df`.

essentially, the following will work, and show what i mean (the comment is there as i don't think tripple-comparisons are allowed in python)

``` python
assert id(pred_df) == id(hist_preds_df) # == self.dd.historic_predictions[pair]
```

i assume it's intentional to FIRST assign it before mutating (won't this be problematic with multi-threaded accesses, where you could read this dataframe at any point within this function?) - but it's still the same object that's being mutated, so using pred_df throughout the function should be fine, too.

Reply:
Sounds like you have an idea of how you want this function to look - go ahead and make your changes. 

This function is called once in the entire lifetime of a single `identifier`.